File: lena-config/src/test/java/org/riversoforion/lena/config/resolvers/PropertyNameResolverTest.java
/*
 * Copyright (c) 2024-2025. Eric McIntyre / Rivers of Orion
 */
package org.riversoforion.lena.config.resolvers;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;
import org.riversoforion.lena.config.Name;
import org.riversoforion.lena.config.Namespace;

import java.io.IOException;
import java.io.InputStream;
import java.util.Properties;

import static org.assertj.core.api.Assertions.assertThat;

class PropertyNameResolverTest {

    @ParameterizedTest(name = "{0}+{1} -> {2}")
    @CsvSource(textBlock = """
                           /, myProp, myProp
                           /, my.otherProp, my.otherProp
                           /, prop-name, prop-name
                           /my-prefix, myProp, my-prefix.myProp
                           /my/prefix, other-prop, my.prefix.other-prop
                           /my/prefix, other.prop, my.prefix.other.prop
                           /, name+with#funky?Characters!, name+with#funky?Characters!
                           /My Prefix, %some$Other@PROP, My Prefix.%some$Other@PROP
                           /Another/Prefix, .valid, Another.Prefix..valid
                           /, .how.about.this?, .how.about.this?
                           /, ., .
                           /It $eem$, property..names.are...Very\tFlexible, It $eem$.property..names.are...Very\tFlexible
                           /, #Not.a.Comment, #Not.a.Comment
                           """)
    @DisplayName("resolveName with valid scenarios")
    void resolveName_Valid(String namespace, String rawName, String expected) throws IOException {

        Properties testProps = new Properties();
        testProps.load(testProperties());
        Namespace ns = Namespace.parse(namespace);
        Name name = Name.of(rawName);
        PropertyNameResolver resolver = new PropertyNameResolver();

        String result = resolver.resolveName(ns, name);

        assertThat(result).isEqualTo(expected);
        // Make sure that property names we think are valid can actually be used to look up a property
        assertThat(testProps).containsKey(expected);
    }

    private InputStream testProperties() {

        return this.getClass().getResourceAsStream("/properties/prop-name-resolver.properties");
    }
}

File: lena-config/src/jvmTest/kotlin/org/riversoforion/lena/config/resolvers/PropertyNameResolverTest.kt
/*
 * Copyright (c) 2024-2026. Eric McIntyre / Rivers of Orion
 */
package org.riversoforion.lena.config.resolvers

import org.riversoforion.lena.config.Name
import org.riversoforion.lena.config.Namespace
import java.util.Properties
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class PropertyNameResolverTest {

    @Test
    fun resolveName_Valid() {
        val cases = listOf(
            Triple("/", "myProp", "myProp"),
            Triple("/", "my.otherProp", "my.otherProp"),
            Triple("/", "prop-name", "prop-name"),
            Triple("/my-prefix", "myProp", "my-prefix.myProp"),
            Triple("/my/prefix", "other-prop", "my.prefix.other-prop"),
            Triple("/my/prefix", "other.prop", "my.prefix.other.prop"),
            Triple("/", "name+with#funky?Characters!", "name+with#funky?Characters!"),
            Triple("/My Prefix", "%some\$Other@PROP", "My Prefix.%some\$Other@PROP"),
            Triple("/Another/Prefix", ".valid", "Another.Prefix..valid"),
            Triple("/", ".how.about.this?", ".how.about.this?"),
            Triple("/", ".", "."),
            Triple("/It \$eem\$", "property..names.are...Very\tFlexible", "It \$eem\$.property..names.are...Very\tFlexible"),
            Triple("/", "#Not.a.Comment", "#Not.a.Comment"),
        )
        val testProps = Properties()
        javaClass.getResourceAsStream("/properties/prop-name-resolver.properties").use { testProps.load(it) }
        val resolver = PropertyNameResolver()

        for ((namespace, rawName, expected) in cases) {
            val ns = Namespace.parse(namespace)
            val name = Name.of(rawName)

            val result = resolver.resolveName(ns, name)

            assertEquals(expected, result, "namespace='$namespace' rawName='$rawName'")
            // Make sure that property names we think are valid can actually be used to look up a property
            assertTrue(testProps.containsKey(expected), "expected properties file to contain key '$expected'")
        }
    }
}
